ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed

ccl/multitenantccl/tenantcostclient.TestScheduledJobsConsumption [failed](https://tanzanite.cluster.engflow.com/invocations/default/d1d23441-cf8e-40bd-a195-2a5c3518f525?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpdGVuYW50Y2NsL3RlbmFudGNvc3RjbGllbnQ6dGVuYW50Y29zdGNsaWVudF90ZXN0) on release-24.1 @ [6e3e9e6012dc04dabc64333a87e77a80d9f9d46b](https://github.com/cockroachdb/cockroach/commits/6e3e9e6012dc04dabc64333a87e77a80d9f9d46b):


```
=== RUN   TestScheduledJobsConsumption
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestScheduledJobsConsumption1000854501
    test_log_scope.go:81: use -show-logs to present logs inline
    tenant_side_test.go:1148: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/multitenantccl/tenantcostclient_test/pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go:1148
        	Error:      	"12" is not less than or equal to "3"
        	Test:       	TestScheduledJobsConsumption
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestScheduledJobsConsumption1000854501
--- FAIL: TestScheduledJobsConsumption (20.52s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #122333 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.2 release-blocker]
- #122331 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.1 release-blocker]
</p>
</details>
/cc @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestScheduledJobsConsumption.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-37810

ccl/multitenantccl/tenantcostclient.TestScheduledJobsConsumption [failed](https://tanzanite.cluster.engflow.com/invocations/default/03df0176-eeda-4513-a7fa-0926b01d41cc?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpdGVuYW50Y2NsL3RlbmFudGNvc3RjbGllbnQ6dGVuYW50Y29zdGNsaWVudF90ZXN0) on release-24.1 @ [273a97e181d634e5d5d51ac7d3e05f3f4f31887a](https://github.com/cockroachdb/cockroach/commits/273a97e181d634e5d5d51ac7d3e05f3f4f31887a):


```
=== RUN   TestScheduledJobsConsumption
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestScheduledJobsConsumption700134005
    test_log_scope.go:81: use -show-logs to present logs inline
    tenant_side_test.go:1148: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/multitenantccl/tenantcostclient_test/pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go:1148
        	Error:      	"5" is not less than or equal to "3"
        	Test:       	TestScheduledJobsConsumption
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestScheduledJobsConsumption700134005
--- FAIL: TestScheduledJobsConsumption (12.39s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #122592 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.2 release-blocker]
- #122333 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.2 release-blocker]
- #122331 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.1 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestScheduledJobsConsumption.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/multitenantccl/tenantcostclient.TestScheduledJobsConsumption [failed](https://tanzanite.cluster.engflow.com/invocations/default/d355363f-bde6-4e8e-9de4-9fdce8b8e161?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpdGVuYW50Y2NsL3RlbmFudGNvc3RjbGllbnQ6dGVuYW50Y29zdGNsaWVudF90ZXN0) on release-24.1 @ [88c9d88cb0093d35f2d630137fd178518aa48569](https://github.com/cockroachdb/cockroach/commits/88c9d88cb0093d35f2d630137fd178518aa48569):


```
=== RUN   TestScheduledJobsConsumption
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestScheduledJobsConsumption3454015885
    test_log_scope.go:81: use -show-logs to present logs inline
    tenant_side_test.go:1148: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/multitenantccl/tenantcostclient_test/pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go:1148
        	Error:      	"21" is not less than or equal to "3"
        	Test:       	TestScheduledJobsConsumption
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestScheduledJobsConsumption3454015885
--- FAIL: TestScheduledJobsConsumption (20.21s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #122740 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.2.5-rc release-blocker]
- #122739 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.1.20-rc release-blocker]
- #122592 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.2 release-blocker]
- #122333 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.2 release-blocker]
- #122331 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.1 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestScheduledJobsConsumption.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/multitenantccl/tenantcostclient.TestScheduledJobsConsumption [failed](https://tanzanite.cluster.engflow.com/invocations/default/a5335b8c-832c-48be-9121-364ca32ef1be?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpdGVuYW50Y2NsL3RlbmFudGNvc3RjbGllbnQ6dGVuYW50Y29zdGNsaWVudF90ZXN0) on release-24.1 @ [3f3127b8b592b9c63df76e47cb9f0a3b47357f62](https://github.com/cockroachdb/cockroach/commits/3f3127b8b592b9c63df76e47cb9f0a3b47357f62):


```
=== RUN   TestScheduledJobsConsumption
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestScheduledJobsConsumption732571998
    test_log_scope.go:81: use -show-logs to present logs inline
    tenant_side_test.go:1148: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/multitenantccl/tenantcostclient_test/pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go:1148
        	Error:      	"8" is not less than or equal to "3"
        	Test:       	TestScheduledJobsConsumption
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestScheduledJobsConsumption732571998
--- FAIL: TestScheduledJobsConsumption (21.71s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #122740 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.2.5-rc]
- #122739 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.1.21-rc]
- #122592 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.2]
- #122333 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.2]
- #122331 ccl/multitenantccl/tenantcostclient: TestScheduledJobsConsumption failed [C-test-failure O-robot T-multitenant branch-release-23.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestScheduledJobsConsumption.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/base/testing_knobs.go
+++ b/pkg/base/testing_knobs.go
@@ -58,4 +58,5 @@ type TestingKnobs struct {
 	LOQRecovery                    ModuleTestingKnobs
 	KeyVisualizer                  ModuleTestingKnobs
 	TenantCapabilitiesTestingKnobs ModuleTestingKnobs
+	TableStatsKnobs                ModuleTestingKnobs
 }

--- a/pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go
+++ b/pkg/ccl/multitenantccl/tenantcostclient/tenant_side_test.go
@@ -1123,6 +1123,9 @@ func TestScheduledJobsConsumption(t *testing.T) {
 					RetryMaxDelay:     &zeroDuration,
 				},
 			},
+			TableStatsKnobs: &stats.TableStatsTestingKnobs{
+				DisableInitialTableCollection: true,
+			},
 		},
 	})
 
@@ -1142,10 +1145,9 @@ func TestScheduledJobsConsumption(t *testing.T) {
 	after.Sub(&before)
 	require.Zero(t, after.WriteBatches)
 	require.Zero(t, after.WriteBytes)
-	// Expect up to 3 batches for initial auto-stats query, schema catalog fill,
-	// and anything else that happens once during server startup but might not be
-	// done by this point.
-	require.LessOrEqual(t, after.ReadBatches, uint64(3))
+	// Expect up to 2 batches for schema catalog fill and anything else that
+	// happens once during server startup but might not be done by this point.
+	require.LessOrEqual(t, after.ReadBatches, uint64(2))
 
 	// Make sure that at least 100 writes (deletes) are reported. The TTL job
 	// should not be exempt from cost control.

--- a/pkg/server/server_sql.go
+++ b/pkg/server/server_sql.go
@@ -1107,6 +1107,10 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 	if externalConnKnobs := cfg.TestingKnobs.ExternalConnection; externalConnKnobs != nil {
 		execCfg.ExternalConnectionTestingKnobs = externalConnKnobs.(*externalconn.TestingKnobs)
 	}
+	var tableStatsTestingKnobs *stats.TableStatsTestingKnobs
+	if tableStatsKnobs := cfg.TestingKnobs.TableStatsKnobs; tableStatsKnobs != nil {
+		tableStatsTestingKnobs = tableStatsKnobs.(*stats.TableStatsTestingKnobs)
+	}
 
 	// Set up internal memory metrics for use by internal SQL executors.
 	// Don't add them to the registry now because it will be added as part of pgServer metrics.
@@ -1160,6 +1164,7 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 		internalDB,
 		execCfg.TableStatsCache,
 		stats.DefaultAsOfTime,
+		tableStatsTestingKnobs,
 	)
 	execCfg.StatsRefresher = statsRefresher
 	distSQLServer.ServerConfig.StatsRefresher = statsRefresher

--- a/pkg/sql/stats/BUILD.bazel
+++ b/pkg/sql/stats/BUILD.bazel
@@ -22,6 +22,7 @@ go_library(
     importpath = "github.com/cockroachdb/cockroach/pkg/sql/stats",
     visibility = ["//visibility:public"],
     deps = [
+        "//pkg/base",
         "//pkg/clusterversion",
         "//pkg/jobs/jobspb",
         "//pkg/keys",

--- a/pkg/sql/stats/automatic_stats.go
+++ b/pkg/sql/stats/automatic_stats.go
@@ -18,6 +18,7 @@ import (
 	"sync"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
@@ -219,6 +220,7 @@ type Refresher struct {
 	internalDB descs.DB
 	cache      *TableStatisticsCache
 	randGen    autoStatsRand
+	knobs      *TableStatsTestingKnobs
 
 	// mutations is the buffered channel used to pass messages containing
 	// metadata about SQL mutations to the background Refresher thread.
@@ -277,6 +279,17 @@ type settingOverride struct {
 	settings catpb.AutoStatsSettings
 }
 
+// TableStatsTestingKnobs contains testing knobs for table statistics.
+type TableStatsTestingKnobs struct {
+	// DisableInitialTableCollection, if set, indicates that the "initial table
+	// collection" performed by the Refresher should be skipped.
+	DisableInitialTableCollection bool
+}
+
+var _ base.ModuleTestingKnobs = &TableStatsTestingKnobs{}
+
+func (k *TableStatsTestingKnobs) ModuleTestingKnobs() {}
+
 // MakeRefresher creates a new Refresher.
 func MakeRefresher(
 	ambientCtx log.AmbientContext,
@@ -284,6 +297,7 @@ func MakeRefresher(
 	internalDB descs.DB,
 	cache *TableStatisticsCache,
 	asOfTime time.Duration,
+	knobs *TableStatsTestingKnobs,
 ) *Refresher {
 	randSource := rand.NewSource(rand.Int63())
 
@@ -293,6 +307,7 @@ func MakeRefresher(
 		internalDB:       internalDB,
 		cache:            cache,
 		randGen:          makeAutoStatsRand(randSource),
+		knobs:            knobs,
 		mutations:        make(chan mutation, refreshChanBufferLen),
 		settings:         make(chan settingOverride, refreshChanBufferLen),
 		asOfTime:         asOfTime,
@@ -428,6 +443,9 @@ func (r *Refresher) Start(
 		for {
 			select {
 			case <-initialTableCollection:
+				if r.knobs != nil && r.knobs.DisableInitialTableCollection {
+					continue
+				}
 				r.ensureAllTables(ctx, initialTableCollectionDelay)
 				if len(r.mutationCounts) > 0 {
 					ensuringAllTables = true

--- a/pkg/sql/stats/automatic_stats_test.go
+++ b/pkg/sql/stats/automatic_stats_test.go
@@ -73,7 +73,7 @@ func TestMaybeRefreshStats(t *testing.T) {
 		s.InternalDB().(descs.DB),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
-	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */)
+	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
 
 	// There should not be any stats yet.
 	if err := checkStatsCount(ctx, cache, descA, 0 /* expected */); err != nil {
@@ -197,7 +197,7 @@ func TestEnsureAllTablesQueries(t *testing.T) {
 		s.InternalDB().(descs.DB),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
-	r := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */)
+	r := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
 
 	// Exclude the 3 system tables which don't use autostats.
 	systemTablesWithStats := bootstrap.NumSystemTablesForSystemTenant - 3
@@ -299,7 +299,7 @@ func BenchmarkEnsureAllTables(b *testing.B) {
 				s.InternalDB().(descs.DB),
 			)
 			require.NoError(b, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
-			r := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */)
+			r := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
 
 			b.ResetTimer()
 			for i := 0; i < b.N; i++ {
@@ -372,7 +372,7 @@ func TestAverageRefreshTime(t *testing.T) {
 		s.InternalDB().(descs.DB),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
-	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */)
+	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
 
 	// curTime is used as the current time throughout the test to ensure that the
 	// calculated average refresh time is consistent even if there are delays due
@@ -619,7 +619,7 @@ func TestAutoStatsReadOnlyTables(t *testing.T) {
 		s.InternalDB().(descs.DB),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
-	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */)
+	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
 
 	AutomaticStatisticsClusterMode.Override(ctx, &st.SV, true)
 
@@ -674,7 +674,7 @@ func TestAutoStatsOnStartupClusterSettingOff(t *testing.T) {
 		s.InternalDB().(descs.DB),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
-	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */)
+	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
 
 	// Refresher start should trigger stats collection on t.a.
 	if err := refresher.Start(
@@ -721,7 +721,7 @@ func TestNoRetryOnFailure(t *testing.T) {
 		s.InternalDB().(descs.DB),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
-	r := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */)
+	r := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
 
 	// Try to refresh stats on a table that doesn't exist.
 	r.maybeRefreshStats(

